Improve Player Lifecycle Docs (#869)
spatialos/gdk-for-unity: docs/modules/player-lifecycle/heartbeating.md

<%(TOC)%>
# Heartbeating

> Heartbeating is how a server-worker continually checks that a client-worker has not disconnected from SpatialOS.

Reviewer: Should explain the motivation for doing this?

## How does it work?

> Generally, a server sends a heartbeat request at regular intervals to clients. If the server continually receives unsuccessful responses back from a client, the server will assume the client has died. The server then proceeds to delete the Player associated with that client.

Reviewer: ~unsuccessful responses~ --> failures, no response, etc?

Reviewer: The server then ~proceeds to~ delete ...

## How is it implemented in the GDK?

### Initialisation

> On the server-worker, the [`PlayerHeartbeatInitializationSystem`](https://github.com/spatialos/gdk-for-unity/blob/master/workers/unity/Packages/com.improbable.gdk.playerlifecycle/Systems/PlayerHeartbeat/PlayerHeartbeatInitializationSystem.cs) continually checks for ECS entities in the worker's view that have a `PlayerHeartbeatClient` component, but do not yet have the `HeartbeatData` component. For all these entities, a `HeartbeatData` component is added to the ECS entity at the end of the current tick.

Reviewer: ~continually~

> The `HeartbeatData` component is used to determine what entities need to be sent `PlayerHeartbeat` requests, and also to track how many consecutive failed heartbeats a player has.

Reviewer: used to determine ~what~ **which** entities

Reviewer: and ~also~ to track

Reviewer: No need for the comma before the and. The list only contains two items

### Sending PlayerHeartbeat requests

> At intervals set by `PlayerLifecycleConfig.PlayerHeartbeatIntervalSeconds`, the [`SendPlayerHeartbeatRequestSystem`](https://github.com/spatialos/gdk-for-unity/blob/master/workers/unity/Packages/com.improbable.gdk.playerlifecycle/Systems/PlayerHeartbeat/SendPlayerHeartbeatRequestSystem.cs) finds all entities where the server-worker has write authority over the `PlayerHeartbeatServer` component and the entity contains this new `HeartbeatData` component. It then sends a `PlayerHeartbeat` request to each entity, which is expected to by handled by the client-worker.

Reviewer: ```contains this new `HeartbeatData` component```

to

```has the `HeartbeatData` component```

<%(TOC)%>
# Heartbeating

> Heartbeating is a technique used to regularly checking-in on the health of connected client-workers and ensuring that unresponsive or disconnected clients are periodically removed from the game world.

Reviewer 2: ```suggestion
Heartbeating is a technique used to ensure that client-workers are still connected. Unresponsive or disconnected clients are periodically removed from the game world.
```

> ## Why is it useful?

Reviewer 2: I thought we don't use question marks for headers? @jamiebrynes7

Reviewer: I don't think we do

> Client-workers typically create a new Player entity at some point after connecting to SpatialOS. Without the heartbeats, this Player entity would simply exist in the world forever as client-workers would create a different, new Player entity on each new connection.

Reviewer 2: ```suggestion
Client-workers typically create a new player entity at some point after connecting to SpatialOS. Without the heartbeats, it is impossible to tell for a server-worker whether the client-worker is still connected and responsive.
```

> By introducing heartbeats, we can ensure that a Player entity corresponding to a disconnected or unresponsive client-worker is deleted from the world. This assures us that, in a stable deployment, the number of Player entities in a world is equal to the number of client-workers connected.

Reviewer 2: ```suggestion
By introducing heartbeats, we ensure that a player entity belonging to a disconnected or unresponsive client-worker is deleted from the world. This assures us that, in a stable deployment, the number of Player entities in a world is equal to the number of client-workers connected.
```

## How does it work?

> Generally, a server sends a heartbeat request at regular intervals to clients. If the server continually receives failures or is unable to get a response back from a client, the server will assume the client has disconnected. The server then deletes the Player associated with that client.

Reviewer 2: ```suggestion
A server sends a heartbeat request at regular intervals to the clients. If the server continually receives failures or the request times out, the server assumes that the client has disconnected. The server then deletes the player associated with that client.
```

## How is it implemented in the GDK?

### Initialisation

> On the server-worker, the [`PlayerHeartbeatInitializationSystem`](https://github.com/spatialos/gdk-for-unity/blob/master/workers/unity/Packages/com.improbable.gdk.playerlifecycle/Systems/PlayerHeartbeat/PlayerHeartbeatInitializationSystem.cs) checks for ECS entities in the worker's view that have a `PlayerHeartbeatClient` component, but do not yet have the `HeartbeatData` component. For all these entities, a `HeartbeatData` component is added to the ECS entity at the end of the current tick.

Reviewer 2: do we ever define what a tick is? I vaguely remember having a chat with @ElleEss  months ago about not introducing the word tick but using update or frame. would need to check in the docs though

The `HeartbeatData` component is used to determine which entities need to be sent `PlayerHeartbeat` requests and to track how many consecutive failed heartbeats a player has.

### Sending PlayerHeartbeat requests

> At intervals set by `PlayerLifecycleConfig.PlayerHeartbeatIntervalSeconds`, the [`SendPlayerHeartbeatRequestSystem`](https://github.com/spatialos/gdk-for-unity/blob/master/workers/unity/Packages/com.improbable.gdk.playerlifecycle/Systems/PlayerHeartbeat/SendPlayerHeartbeatRequestSystem.cs) finds all entities where the server-worker has write authority over the `PlayerHeartbeatServer` component and the entity has the `HeartbeatData` component. It then sends a `PlayerHeartbeat` request to each entity, which is expected to by handled by the client-worker.

Reviewer 2: ```suggestion
At intervals set by `PlayerLifecycleConfig.PlayerHeartbeatIntervalSeconds`, the [`SendPlayerHeartbeatRequestSystem`](https://github.com/spatialos/gdk-for-unity/blob/master/workers/unity/Packages/com.improbable.gdk.playerlifecycle/Systems/PlayerHeartbeat/SendPlayerHeartbeatRequestSystem.cs) finds all entities fulfilling the following constraints:
* the server-worker has authority over the `PlayerHeartbeatServer` component
* the entity has the `HeartbeatData` component. 

It sends a `PlayerHeartbeat` request to each of those entities. These requests need to be handled by the client-worker that the player entity belongs to.
```

> Note that there are numerous ways that the request may fail, as outlined [here](https://docs.improbable.io/reference/latest/shared/design/commands#failure-modes).

Reviewer 2: ```suggestion
Note that there are numerous ways that a request may fail, as outlined [here](https://docs.improbable.io/reference/latest/shared/design/commands#failure-modes).
```

> ### Handling PlayerHeartbeat requests

Reviewer 2: same comment
```suggestion
### How to handle `PlayerHeartbeat` requests
```

> The client-worker runs the [`HandlePlayerHeartbeatRequestSystem`](https://github.com/spatialos/gdk-for-unity/blob/master/workers/unity/Packages/com.improbable.gdk.playerlifecycle/Systems/PlayerHeartbeat/HandlePlayerHeartbeatRequestSystem.cs), which sends a response back to each `PlayerHeartbeat` request that was sent to that specific client-worker.

Reviewer 2: ```suggestion
The client-worker runs the [`HandlePlayerHeartbeatRequestSystem`](https://github.com/spatialos/gdk-for-unity/blob/master/workers/unity/Packages/com.improbable.gdk.playerlifecycle/Systems/PlayerHeartbeat/HandlePlayerHeartbeatRequestSystem.cs), which sends a response back whenever it receives a `PlayerHeartbeat` request.
```

### Handling PlayerHeartbeat responses

> Back on the server-worker, the [`HandlePlayerHeartbeatResponseSystem`](https://github.com/spatialos/gdk-for-unity/blob/master/workers/unity/Packages/com.improbable.gdk.playerlifecycle/Systems/PlayerHeartbeat/HandlePlayerHeartbeatResponseSystem.cs) iterates through the `PlayerHeartbeat` responses received. If at least one successful response was received from a particular Player entity, the `NumFailedHeartbeats` inside its `HeartbeatData` component is set to 0.

Reviewer 2: ```suggestion
The [`HandlePlayerHeartbeatResponseSystem`](https://github.com/spatialos/gdk-for-unity/blob/master/workers/unity/Packages/com.improbable.gdk.playerlifecycle/Systems/PlayerHeartbeat/HandlePlayerHeartbeatResponseSystem.cs) on the server-worker iterates through all`PlayerHeartbeat` responses received. If at least one successful response was received from a payer entity, the `NumFailedHeartbeats` inside this player entity's `HeartbeatData` component is set to 0.
```

> If no responses were received with a `Success` status code from a Player entity, the `NumFailedHeartbeats` field is incremented. In the case that an entity has more failed heartbeats than the number configured in `PlayerLifecycleConfig.MaxNumFailedPlayerHeartbeats`, a request to delete the SpatialOS entity is sent.

Reviewer 2: ```suggestion
If no responses were received with a `Success` status code from a Player entity, the `NumFailedHeartbeats` field is incremented. If  an entity has more failed heartbeats than the number configured in `PlayerLifecycleConfig.MaxNumFailedPlayerHeartbeats`, a request to delete the SpatialOS entity is sent to the SpatialOS Runtime.
```

<%(TOC)%>
# Heartbeating

Heartbeating is a technique used to regularly checking-in on the health of connected client-workers and ensuring that unresponsive or disconnected clients are periodically removed from the game world.

## Why it's useful

Client-workers typically create a new Player entity at some point after connecting to SpatialOS. Without the heartbeats, this Player entity would simply exist in the world forever as client-workers would create a different, new Player entity on each new connection.

> By introducing heartbeats, we can ensure that a Player entity corresponding to a disconnected or unresponsive client-worker is deleted from the world. This assures us that, in a stable deployment, the number of Player entities in a world is equal to the number of client-workers connected.

Reviewer: You can have multiple player entities per client. I think the invariant is better expressed as: each Player entity in the world has a client-worker that "owns" it connected or has been connected in the last X seconds. 

Ofc this then brings into question ownership...

## How it works

A server sends a heartbeat request at regular intervals to the clients. If the server continually receives failures or the request times out, the server assumes that the client has disconnected. The server then deletes the player associated with that client.

## How it's implemented in the GDK

### Initialisation

On the server-worker, the [`PlayerHeartbeatInitializationSystem`](https://github.com/spatialos/gdk-for-unity/blob/master/workers/unity/Packages/com.improbable.gdk.playerlifecycle/Systems/PlayerHeartbeat/PlayerHeartbeatInitializationSystem.cs) checks for ECS entities in the worker's view that have a `PlayerHeartbeatClient` component, but do not yet have the `HeartbeatData` component. For all these entities, a `HeartbeatData` component is added to the ECS entity at the end of the current frame.

The `HeartbeatData` component is used to determine which entities need to be sent `PlayerHeartbeat` requests and to track how many consecutive failed heartbeats a player has.

> ### Sending PlayerHeartbeat requests

Reviewer 2: ```suggestion
### How to send PlayerHeartbeat requests
```

At intervals set by `PlayerLifecycleConfig.PlayerHeartbeatIntervalSeconds`, the [`SendPlayerHeartbeatRequestSystem`](https://github.com/spatialos/gdk-for-unity/blob/master/workers/unity/Packages/com.improbable.gdk.playerlifecycle/Systems/PlayerHeartbeat/SendPlayerHeartbeatRequestSystem.cs) finds all entities fulfilling the following constraints:

* the server-worker has authority over the `PlayerHeartbeatServer` component.
* the entity has the `HeartbeatData` component.

It sends a `PlayerHeartbeat` request to each of those entities. These requests need to be handled by the client-worker that the player entity belongs to.

Note that there are numerous ways that a request may fail, as outlined [here](https://docs.improbable.io/reference/latest/shared/design/commands#failure-modes).

### How to handle PlayerHeartbeat requests

The client-worker runs the [`HandlePlayerHeartbeatRequestSystem`](https://github.com/spatialos/gdk-for-unity/blob/master/workers/unity/Packages/com.improbable.gdk.playerlifecycle/Systems/PlayerHeartbeat/HandlePlayerHeartbeatRequestSystem.cs), which sends a response back whenever it receives a `PlayerHeartbeat` request.

> ### Handling PlayerHeartbeat responses

Reviewer 2: ```suggestion
### How to handle PlayerHeartbeat responses
```
